### DISCUSSION TITLE
Should @useDependency on individual API versions catch version mismatch errors?

### DISCUSSION THREAD

**Discussion by grepory** [microsoft/typespec#3919] [Q&A] [answered]
I think my title is a little misleading, and I might update it, but the gist of this is that I don't understand versioning. :)

In this first example -- [Playground Link](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvdmVyc2lvbmluZyI7Cgp1c2luZyBUeXBlU3BlYy5WyR07CgpAc2VydmljZQpAdXNlRGVwZW5kZW5jeSjKC8gwcy52MSkKbmFtZXNwYWNlIEFwaSB7CgogIG1vZGVsIFN1Y2NlZWRzIHsKIMUBLi4uy0VGcm9tVjEKICB9xDMvLyBJIGV4cGVjdGVkIHRoaXMgdG8gZXJyb3IuylJob3VsZEZhaWxXaG9sedxZMtNZyVd3b3JrIGFuZCBpbmNsdWRlIFRlc3QzRmllbGTkAJHMbVBhcnRpYWxseecAztx0Mcd2fQoKQOcBcGVkKOgBLewBKso6x1BlbnVtIMgqyWQgIHYxLMoMMsYM5QDXICBAYWRky1%2FlAYzEGOYAvuYAkctK5ADjMeYA4zogc3Ry5QHt3E8y0U8yz08y30%2F%2FAJ4xzVHYMcht5wGfym3aOucAus46MvAAp30KCg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D) -- I have an unversioned API that depends on a dependency at `v1`. There are two examples that I expect to be impacted by `@useDependency`. `ShouldFailWholly` should fail (or at least yield an empty model), because it depends on a model that's from `v2`. Then I have a model `ShouldPartiallySucceed` that depends on a model with fields from v1 and v2. I'd expect the field from v2 not to be in the OpenAPI spec.

In the next example -- [Playground Link](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvdmVyc2lvbmluZyI7Cgp1c2luZyBUeXBlU3BlYy5WyR07CgpAc2VydmljZQpAxzRlZCjHIXMpCm5hbWVzcGFjZSDHFGVkIHsKICBlbnVtyBNzxRIgIEB1c2VEZXBlbmRlbmN5KMoLyHFzLnYxKcUrdjEs3zPKMzLHMzLEM30KCiAgQGFkZOsApsdWbW9kZWzqAKdNxA8x5wCbLi4uy1xGcm9txygyxVR9Cv8A%2FiDKPvkA%2F%2BgA1PwAqeUAzOYAqewAhcdCRmllbGQxOiBzdHLlAZzERn0%3D&e=%40typespec%2Fopenapi3&options=%7B%7D) -- I've added useDependency annotations to individual API versions in a versioned API that depends on another versioned API. In this example, I've again made a model that depends on a model from the wrong version, and this time I don't get any error at all.

This last [Playground Link](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvdmVyc2lvbmluZyI7Cgp1c2luZyBUeXBlU3BlYy5WyR07CgpAc2VydmljZQpAxzRlZCjHIXMpCm5hbWVzcGFjZSDHFGVkIHsKICBlbnVtyBNzxRIgIEB1c2VEZXBlbmRlbmN5KMoLyHFzLnYxKcUrdjEsxggyxAh9CgogIC8vIEkgZXhwZWN0IHRoaXMgdG8gc3VjY2VlZCBhbmQgaXQgZG9lcy4KICBAYWRk6wCmx1Ztb2RlbCBTxjPoAJQgIC4uLusAilRlc3QxzXN3b3VsZNB5ZmFpbCBlbnRpcmVsed9zIFNoxENGYWlsV2hvbGx513tGcm9tx0Ay6wCCVMRzd29ya3MgYXPnAIhlZC4g5ACqM0ZpZWxkMSBpcyBpbmNsdWRlZCwgYnV0yx0yxB1ub3T%2FAKUgQmF6%2FACYMVbmAJR9Cv8CCCDKOsdQ9AIL7AHk6gHo5wD1%2BACp6ACr5QGky0nFEOYBCjogc3Ry5QK43E4yzE7sAXfPVTLfVfoAo%2BgBNMtR2DHIbecBzMpt2jrnAMDOOjLwAKd9Cgo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D) is just a combination of the previous two, but again illustrates the behaviors.

I'm just a bit confused about what's supposed to be happening and when I should be seeing errors. Am I doing versioning wrong?

**Comment by timotheeguerin** [comment, marked as answer]
Yeah so I think there is a few issues:
1. The should fail when spreading a model not available in the version used. I think this is basically this issue [#3648](https://github.com/microsoft/typespec/issues/3648)
2. The error you see in the first playground link I think is a separate issue could file a bug for. I think the validation is wrongfully complaining here.

That being said I think you should be very careful about versioning libraries. Versioning spec is one thing but versioning libraries(with decorators) can lead to a very confusing understanding of how things get applied.
We started doing that for Azure libraries and due to the confusion have started moving away.

**Reply by grepory** [reply]
Good to know.

I think I'm struggling with versioning in general. I keep running across issues that I really can't explain.
